doc/man/man2/poll.2.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the section entitled "GNU Free
.\" Documentation License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaning to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF
.\" SUCH CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log$
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore T
.R2
.so openss7.macros
.\"
.\"
.TH POLL 2 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Calls"
.\"
.\"
.SH NAME
.B poll
\- wait for some event on a \fISTREAMS\fR file descriptor
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/poll.h>
.HP
.BI "int poll(struct pollfd *" ufds ", unsigned int " nfds ", int " timeout );
.\"
.\"
.SH DESCRIPTION
.PP
.\"
.\"
.SH "RETURN VALUES"
.PP
.\"
.\"
.SH ERRORS
.PP
.TP
.RB [ EBADF ]
.TP
.RB [ ENOMEM ]
.TP
.RB [ EFAULT ]
.TP
.RB [ EINTR ]
.\"
.\"
.SH CAVEATS
.PP
.\"
.\"
.SH BUGS
.PP
.BR poll ()
has no known bugs.
.\"
.\"
.SH EXAMPLES
.PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR streamio (7).
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.PP
.BR poll ()
first appeared in SVR 3.
.[
svr3
.]
.\"
.\"
.XX
.[
magic
.]
.[
svr42
.]
.[
svr4
.]
.[
svr3
.]
.[
svid
.]
.[
xns
.]
.[
xbd5
.]
.[
$LIST$
.]
.TA


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-04-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel20081029_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.

   - Modifications to build under `Fedora' `2.6.22.5-49' kernel.  These
     changes also support `2.6.22.9-91.fc7' kernel.  Modifications to
     build under `Fedora' `2.6.25-45.fc9' and `2.6.26.5-45.fc9' kernels.
     Documented `lib32gcc1' problem on Ubuntu.  Noted problem running
     under XEN kernels.  XEN kernels are not yet supported.  Added
     `MODULE_VERSION' to all modules and drivers.

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.  Improvements to common build
     process for embedded and cross-compile targets.  Cross-compile
     fixes (strap out `AC_FUNC_REALLOC' macro when cross-compiling).
     Conversion of RPM spec files to common approach for major
     subpackages.

     Build system now builds `yum(8)' repositories for RPMs and
     `apt-get(8)' repositories for DEBs.  Installation documentation
     has been updated to include details of repository install
     sourcesref.

   - Higher performance and updated performance papers.

   - Updated tool chain to `m4-1.4.12', `autoconf-2.63' and
     `texinfo-4.13'.  Support for `flex 2.5.33' in maintainer mode.

   - Updated references database for manual pages and roff documents.

   - Added the following major sub-packages to the master build (and
     release):


LiS/head/poll.c
/*****************************************************************************

 @(#) $RCSfile: poll.c,v $ $Name:  $($Revision: 1.1.1.2.4.8 $) $Date: 2008-04-29 08:33:13 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:13 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: poll.c,v $
 Revision 1.1.1.2.4.8  2008-04-29 08:33:13  brian
 - update headers for Affero release

 Revision 1.1.1.2.4.7  2007/08/15 04:58:07  brian
 - GPLv3 updates

 *****************************************************************************/

#ident "@(#) $RCSfile: poll.c,v $ $Name:  $($Revision: 1.1.1.2.4.8 $) $Date: 2008-04-29 08:33:13 $"

static char const ident[] = "$RCSfile: poll.c,v $ $Name:  $($Revision: 1.1.1.2.4.8 $) $Date: 2008-04-29 08:33:13 $";

/************************************************************************
*                      poll() System Call				*
*************************************************************************
*									*
* This file does the hard work for the poll() system call.  It handles	*
* STREAMS files using the AT&T poll mechanism and non-streams files	*
* by using their select functions.					*
*									*
* Author:	David Grothe <dave@gcom.com>				*
*									*
* Copyright (C) 1997  David Grothe, Gcom, Inc <dave@gcom.com>		*
*									*
************************************************************************/

#include <sys/stream.h>
#ifdef LINUX_POLL
#define	USE_LINUX_POLL_H	1
#include <linux/poll.h>
#else
#include <sys/poll.h>
#endif
#include <sys/lislocks.h>
#include <sys/osif.h>

char *lis_poll_file =
#ifdef USED_LINUX_POLL_H
    "<linux/poll.h>"
#else
#ifdef USED_LIS_POLL_H
    "<LiS/include/sys/poll.h>"
#else
    "<unknown/poll.h>"
#endif
#endif
;

/* typedef struct semaphore	semaphore_t ; */
typedef struct file file_t;

long lis_poll_sleeping;			/* # times slept > twice on same poll */

#ifdef POLL_TRACE

typedef struct {
	int time_out;
	long target_time;
	long jiffies;
	long time_interval;
	long ticks;

} poll_trace_t;

poll_trace_t lis_poll_trace[5];

#endif

#ifdef PORTABLE_POLL
/*
 * Routines elsewhere in the Linux kernel
 */
extern void select_free_wait(select_table * p);
extern int select_check(int flag, select_table * wait, struct file *file);
#endif

/*
 * Routines elsewhere in STREAMS
 */
extern char *lis_poll_events(short events);
extern unsigned lis_poll_bits(stdata_t *hd);

/************************************************************************
*                            lis_poll_wakeup                            *
*************************************************************************
*									*
* Called via the polldat structure to wakeup sleeping poll routine.	*
*									*
************************************************************************/
#ifdef PORTABLE_POLL
void
lis_poll_wakeup(long arg)
{
	struct wait_queue **poll_wait_addr_ptr = (struct wait_queue **) arg;

	wake_up(poll_wait_addr_ptr);

}				/* lis_poll_wakeup */
#endif

/************************************************************************
*                            poll_do_select                             *
*************************************************************************
*									*
* This routine checks the file descriptor using the select mechanism.	*
*									*
* We map the different combinations of poll conditions down to the three*
* select conditions and use the Linux select mechanism to test the	*
* file for these conditions.						*
*									*
************************************************************************/
#ifdef PORTABLE_POLL
int
poll_do_select(select_table * wait, file_t * fp, polldat_t * pdat_ptr)
{
	short events = pdat_ptr->pd_events;	/* poll events */
	short revents = 0;
	short msk;

	if ((msk = (events & (POLLIN | POLLPRI | POLLRDNORM | POLLRDBAND))) != 0
	    && select_check(SEL_IN, wait, fp) != 0) {
		revents |= msk;
	}

	if ((msk = (events & (POLLOUT | POLLWRNORM | POLLWRBAND))) != 0
	    && select_check(SEL_OUT, wait, fp) != 0) {
		revents |= msk;
	}

	if ((msk = (events & (POLLMSG | POLLNVAL | POLLERR | POLLHUP))) != 0
	    && select_check(SEL_EX, wait, fp) != 0) {
		revents |= msk;
	}

	if (LIS_DEBUG_POLL) {
		struct inode *i;
		kdev_t rdev;

		i = fp->f_inode;
		if (i != NULL)
			rdev = i->i_rdev;
		else
			rdev = 0;

		printk("poll_do_select:  dev=(%d,%d) events:%s revents:%s\n", MAJOR(rdev),
		       MINOR(rdev), lis_poll_events((short) events),
		       lis_poll_events((short) revents));
	}

	return (revents);

}				/* poll_do_select */
#endif

/************************************************************************
*                             lis_poll                                  *
*************************************************************************
*									*
* Poll file descriptors for operations.					*
*									*
************************************************************************/
#ifdef PORTABLE_POLL
int
lis_poll(struct pollfd *fds, int nfds, int time_out)
{
	volatile int err;
	unsigned long size = nfds * sizeof(pollfd_t);
	long target_time;		/* in ms */
	int timer_id = 0;
	long time_interval = time_out;	/* in ms */
	long ms_per_tick = 1000 / HZ;
	long ticks;			/* in system ticks */
	struct pollfd *pfds;
	struct pollfd *pfd_ptr;
	polldat_t *pdat_head;
	polldat_t *pdat_ptr;
	file_t *fp;
	int i;
	int done;
	int ready_fd_cnt = 0;
	int psw;
	int sleep_cnt;
	struct wait_queue poll_wait = { current, NULL };
	struct wait_queue *poll_wait_address;
	select_table poll_select_wait_table;
	select_table *poll_select_wait = &poll_select_wait_table;
	struct select_table_entry *poll_select_table;
	int select_table_size;
	int is_stream = 0;
	extern struct file_operations lis_streams_fops;

	/* 
	 * Round time interval up to a multiple of system ticks.
	 */
	ticks = (time_interval + ms_per_tick - 1) / ms_per_tick;
	target_time = lis_target_time(ticks * ms_per_tick);

	if ((err = lis_check_umem(NULL, VERIFY_WRITE, fds, size)) < 0
	    || (err = lis_check_umem(NULL, VERIFY_READ, fds, size)) < 0)
		return (err);

	if ((pfds = ALLOCF(size, "poll-fd ")) == NULL)
		return (-ENOMEM);

	if ((pdat_head = ALLOCF(nfds * sizeof(polldat_t), "poll-data ")) == NULL) {
		FREE(pfds);
		return (-ENOMEM);
	}

	select_table_size = nfds * 3 * sizeof(struct select_table_entry);

	if (!(poll_select_table = (struct select_table_entry *) __get_free_page(GFP_KERNEL))) {
		FREE(pfds);
		FREE(pdat_head);
		return -ENOMEM;
	}

	memset(pdat_head, 0, nfds * sizeof(polldat_t));	/* clear to zero */
	memset(poll_select_table, 0, 4096);	/* clear to zero */
	poll_select_wait_table.nr = 0;
	poll_select_wait_table.entry = poll_select_table;

	/* 
	 * Obtain copy of user's poll descriptor list
	 */
	if ((err = lis_copyin(NULL, pfds, fds, size)) < 0) {
		FREE(pfds);
		FREE(pdat_head);
		return (err);
	}

	/* 
	 * Initialize our wait address pointer
	 */
	init_waitqueue(&poll_wait_address);

	/* 
	 * Run through the descriptor list and check each file descriptor
	 * for validity.  Mark the ones that are bad.
	 */
	for (i = 0, pdat_ptr = pdat_head, pfd_ptr = pfds; i < nfds; i++, pdat_ptr++, pfd_ptr++) {
		if (pfd_ptr->fd >= NR_OPEN || (fp = current->files->fd[pfd_ptr->fd]) == NULL || fp->f_inode == NULL) {	/* not 
															   an 
															   open 
															   file 
															   descriptor 
															 */
			pfd_ptr->revents = POLLNVAL;
			continue;
		}

		pfd_ptr->revents = 0;	/* ensure POLLNVAL not set */
		pdat_ptr->pd_fn = lis_poll_wakeup;
		pdat_ptr->pd_arg = (long) &poll_wait_address;
	}

	/* 
	 * Loop until we find a valid file descriptor which statisfies the
	 * required conditions.  If there were no good file descriptors
	 * then just fall through and return to the user.
	 *
	 * If no file descriptors are found then wait.  When something
	 * happens on one of the streams then the lis_poll_wakeup function
	 * will get called, which will wake us up.  We then go around and
	 * poll again.
	 *
	 * When we exit the loop we will unlink the polldat entries and
	 * return to the user.
	 *
	 * The file system poll function is called with the file info and
	 * a pointer to a polldat structure.  The file system routine
	 * interrogates the file for the requested conditions according
	 * to the 'pd_events' field in the polldat structure.  It returns
	 * a bit-mask of the events that satisfy the conditions, if any.
	 * The file system also sets the pd_headp pointer to point to
	 * its list head in its stream structure for the polldat list.
	 * We take care of linking the structure into the list.
	 *
	 */
	add_wait_queue(&poll_wait_address, &poll_wait);
	for (done = 0, sleep_cnt = 0; !done;) {
		current->state = TASK_INTERRUPTIBLE;
		for (i = 0, ready_fd_cnt = 0, pdat_ptr = pdat_head, pfd_ptr = pfds; i < nfds;
		     i++, pdat_ptr++, pfd_ptr++) {
			if (pfd_ptr->fd < 0 || pfd_ptr->revents & POLLNVAL)
				continue;	/* only look at good ones */

			pdat_ptr->pd_events = pfd_ptr->events;
			fp = current->files->fd[pfd_ptr->fd];
/* FIXME - use a test not dependent on lis_streams_fops */
			is_stream = fp->f_op == &lis_streams_fops;
			 /**/ if (is_stream) {
				if ((pfd_ptr->revents =
				     (short) lis_strpoll(fp->f_inode, fp, pdat_ptr)) != 0) {
					done = 1;	/* loop breaker */
					ready_fd_cnt++;	/* another ready descriptor */
					continue;	/* skip list manipulations */
				}
			} else {	/* non-stream file */
				/* use select */
				if ((pfd_ptr->revents =
				     poll_do_select(poll_select_wait, fp, pdat_ptr)) != 0) {
					done = 1;	/* loop breaker */
					ready_fd_cnt++;	/* another ready descriptor */
					poll_select_wait = NULL;	/* no more waiting */
					continue;	/* skip list manipulations */
				}
			}

			if (ready_fd_cnt)	/* skip list manipulations */
				continue;	/* if have ready fds */

			if (!is_stream || pdat_ptr->pd_headp == NULL
			    || pdat_ptr->pd_headp->ph_list == pdat_ptr || pdat_ptr->pd_next != NULL
			    || pdat_ptr->pd_prev != NULL)
				continue;	/* no list or already in it */

			lis_spin_lock_irqsave(&pdat_ptr->pd_headp->sd_lock, &psw);

			if (pdat_ptr->pd_headp->ph_list == NULL) {	/* first entry */
				pdat_ptr->pd_next = pdat_ptr->pd_prev = NULL;
				pdat_ptr->pd_headp->ph_list = pdat_ptr;
			} else {	/* multiple entries */

				pdat_ptr->pd_next = pdat_ptr->pd_headp->ph_list;
				pdat_ptr->pd_prev = NULL;

				pdat_ptr->pd_headp->ph_list->pd_prev = pdat_ptr;
				pdat_ptr->pd_headp->ph_list = pdat_ptr;
			}

			lis_spin_unlock_irqrestore(&pdat_ptr->pd_headp->sd_lock, &psw);
		}

		if (done)	/* done */
			break;	/* exit while loop */

		/* 
		 * If signalled then return to caller
		 */
		if (current->signal & ~current->blocked) {
			err = -EINTR;
			break;
		}

		/* 
		 * We recompute the interval until the original time out
		 * since we may wake up and go around several times before
		 * we finally decide to return.
		 *
		 * A timeout value of -1 means wait forever.
		 */
		if (time_out != -1) {
			if ((time_interval = lis_time_till(target_time)) <= 0)
				break;	/* time's up */

			ticks = (time_interval + ms_per_tick - 1) / ms_per_tick;
			timer_id =
			    timeout((timo_fcn_t *) lis_poll_wakeup, (caddr_t) &poll_wait_address,
				    ticks);
		}
#ifdef POLL_TRACE
		if (sleep_cnt < sizeof(lis_poll_trace) / sizeof(lis_poll_trace[0])) {
			lis_poll_trace[sleep_cnt].time_out = time_out;
			lis_poll_trace[sleep_cnt].target_time = target_time;
			lis_poll_trace[sleep_cnt].jiffies = jiffies;
			lis_poll_trace[sleep_cnt].time_interval = time_interval;
			lis_poll_trace[sleep_cnt].ticks = ticks;
		}
#endif

		if (++sleep_cnt > 2)
			lis_poll_sleeping++;

		schedule();

		if (timer_id != 0)
			untimeout(timer_id);	/* cancel timer */
	}

	current->state = TASK_RUNNING;
	remove_wait_queue(&poll_wait_address, &poll_wait);

	/* 
	 * Go through and unlink all the polldat structures
	 */
	for (i = 0, pdat_ptr = pdat_head; i < nfds; i++, pdat_ptr++) {
		if (pdat_ptr->pd_headp != NULL) {	/* have list head specified */
			lis_spin_lock_irqsave(&pdat_ptr->pd_headp->sd_lock, &psw);

			if (pdat_ptr->pd_headp->ph_list == pdat_ptr)	/* at head */
				pdat_ptr->pd_headp->ph_list = pdat_ptr->pd_next;

			if (pdat_ptr->pd_next)	/* next element exists */
				pdat_ptr->pd_next->pd_prev = pdat_ptr->pd_prev;

			if (pdat_ptr->pd_prev)	/* previous element exists */
				pdat_ptr->pd_prev->pd_next = pdat_ptr->pd_next;

			lis_spin_unlock_irqrestore(&pdat_ptr->pd_headp->sd_lock, &psw);
		}
	}

	FREE(pdat_head);	/* done with polldat structures */

	select_free_wait(&poll_select_wait_table);	/* undo wait q entries */

	free_page((unsigned long) poll_select_table);	/* done with select table */

	if (err == 0)
		err = lis_copyout(NULL, pfds, fds, size);	/* update user's poll list */

	FREE(pfds);		/* done with fd list */

	if (err < 0)
		return (err);	/* error return from semaphore */
	else
		return (ready_fd_cnt);	/* return # of ready fds */

}				/* lis_poll */
#endif

/************************************************************************
*                           lis_poll_2_1                                *
*************************************************************************
*									*
* This is the kernel release 2.1 poll routine.  It is pointed to by	*
* the fops structure for STREAMS.  It handles polling for one stream.	*
* It builds a list of processes waiting on the stream in the task list	*
* owned in the stdata structure.					*
*									*
* If we have to wait then lis_wake_up_poll in head.c does the wakeup.	*
*									*
************************************************************************/
#ifdef LINUX_POLL
unsigned
lis_poll_2_1(struct file *fp, poll_table * wait)
{
	stdata_t *head;

	if (fp == NULL)
		return (POLLNVAL);

	head = FILE_STR(fp);
	if (head->magic != STDATA_MAGIC) {
		printk("lis_poll_2_1: fp=%p wait=%p head=%p magic=%lx should be %lx\n", fp, wait,
		       head, head->magic, STDATA_MAGIC);
		return (POLLNVAL);
	}

	/* 
	 * Put us into the wait queue first.  If any files have returned
	 * non-zero then Linux stops the wait queue insertion.  Linux poll
	 * system call cleans up the wait queues after all the poll operations
	 * are done.
	 */
	poll_wait(fp, &head->sd_task_list, wait);
	return (lis_poll_bits(head));

}				/* lis_poll_2_1 */
#endif
